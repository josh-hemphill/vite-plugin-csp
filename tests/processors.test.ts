import {
	builtinProcessorFns,
	builtinProcessorFiles,
} from '../src/builtin-processors/index.js';
import mockFs from 'mock-fs';
import {
	DebugAble,
	InternalProcessFn,
	InternalProcessFnNames,
	ProcessFnContext,
} from '../src/consts.js';
import { CspDirectiveHeaders } from 'csp-typed-directives';
import { readFileSync } from 'fs';
import { join } from 'path';
import ViteCspPlugin from '../src/index.js';

const baseContext: ProcessFnContext = {
	path: 'string',
	htmlFileName: 'string',
	srvConfDir: '.server_config',
	builtinProcessorFns: builtinProcessorFns,
};
const baseHeaders: CspDirectiveHeaders = {
	'Content-Security-Policy-Report-Only': '',
	'Content-Security-Policy': '',
	'Report-To': '',
	'Referrer-Policy': '',
};

describe('ViteCspPlugin({ processFn:[] })', () => {
	it('filters invalid processors',()=>{
		// eslint-disable-next-line @typescript-eslint/ban-ts-comment
		//@ts-ignore
		const plugin = ViteCspPlugin({ processFn: ['blah blah',{no:'invalid'},12], debugPlugin: true });
		const { processFns } = (<DebugAble>plugin).debugProperties;
		expect(processFns.length).toStrictEqual(0);
	});
	describe('builtin processors', () => {
		const builtins = <[InternalProcessFnNames, InternalProcessFn][]>(
			Object.entries(builtinProcessorFns)
		);
		describe.each(builtins)('%s(processorOptions)', (name, fn) => {
			it('creates default config', () => {
				mockFs({});
				fn({
					processor: name,
					ctx: baseContext,
					parsedHeaders: baseHeaders,
				});
				const processorDefaultFile = builtinProcessorFiles[name];
				const fileText = readFileSync(
					join(baseContext.srvConfDir, processorDefaultFile),
					{ encoding: 'utf-8' },
				);
				mockFs.restore();
				expect(fileText).toMatchSnapshot(name);
			});
			it('creates custom config', () => {
				const srvConfDir = '.server_config';
				const fileName = name + '.conf';
				mockFs({});
				fn({
					processor: name,
					ctx: {
						...baseContext,
						srvConfDir,
					},
					parsedHeaders: baseHeaders,
					outFile: fileName,
				});
				const fileText = readFileSync(
					join(baseContext.srvConfDir, fileName),
					{ encoding: 'utf-8' },
				);
				mockFs.restore();
				expect(fileText).toMatchSnapshot(name);
			});
		});
		describe('plugin.transformIndexHtml() => buildConfigs()', () => {
			it.each(builtins)('creates default config for %s', (name, _fn) => {
				const plugin = ViteCspPlugin({ processFn: [name] });
				mockFs({});
				if (
					plugin &&
					plugin.transformIndexHtml &&
					typeof plugin.transformIndexHtml === 'function'
				) {
					plugin.transformIndexHtml('', { path: '', filename: '' });
				}
				const processorDefaultFile = builtinProcessorFiles[name];
				const fileText = readFileSync(
					join(baseContext.srvConfDir, processorDefaultFile),
					{ encoding: 'utf-8' },
				);
				mockFs.restore();
				expect(fileText).toMatchSnapshot(name);
			});
			it.each(builtins)('creates custom config for %s', (name, _fn) => {
				const srvConfDir = '.server_config';
				const fileName = name + '.conf';
				const plugin = ViteCspPlugin({
					processFn: [
						{
							processor: name,
							outFile: fileName,
						},
					],
					srvConfDir,
				});
				mockFs({});
				if (
					plugin &&
					plugin.transformIndexHtml &&
					typeof plugin.transformIndexHtml === 'function'
				) {
					plugin.transformIndexHtml('', { path: '', filename: '' });
				}
				const fileText = readFileSync(
					join(baseContext.srvConfDir, fileName),
					{ encoding: 'utf-8' },
				);
				mockFs.restore();
				expect(fileText).toMatchSnapshot(name);
			});
		});
	});
	describe('custom processor', () => {
		it('receives consistent params', async () => {
			const srvConfDir = '.server_config';
			const fileName = 'custom.conf';
			let processFnPropSig: unknown[] = [];
			const plugin = ViteCspPlugin({
				processFn: [
					(...args) => {
						processFnPropSig = args;
						return { name: fileName, content: 'hello world' };
					},
				],
				srvConfDir,
			});
			mockFs({});
			if (
				plugin &&
				plugin.transformIndexHtml &&
				typeof plugin.transformIndexHtml === 'function'
			) {
				await plugin.transformIndexHtml('', { path: '', filename: '' });
			}
			const fileText = readFileSync(
				join(baseContext.srvConfDir, fileName),
				{ encoding: 'utf-8' },
			);
			mockFs.restore();
			expect(fileText).toBe('hello world');
			expect(processFnPropSig).toMatchInlineSnapshot(`
			Array [
			  Object {
			    "builtinProcessorFns": Object {
			      "CaddyJSON": [Function],
			      "CaddyJSON_HeadersOnly": [Function],
			      "Caddyfile": [Function],
			      "Caddyfile_HeadersOnly": [Function],
			      "Nginx": [Function],
			      "Nginx_HeadersOnly": [Function],
			    },
			    "htmlFileName": "",
			    "path": "",
			    "srvConfDir": ".server_config",
			  },
			  Object {
			    "Content-Security-Policy": "object-src 'none'; script-src 'unsafe-inline' 'self' 'unsafe-eval'; script-src-attr; style-src 'unsafe-inline' 'self' 'unsafe-eval'; style-src-attr; base-uri 'self';",
			    "Content-Security-Policy-Report-Only": "",
			    "Referrer-Policy": "strict-origin-when-cross-origin",
			    "Report-To": "",
			  },
			]
		`);
		});
	});
});
