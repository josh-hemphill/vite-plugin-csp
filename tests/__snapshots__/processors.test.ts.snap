// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViteCspPlugin({ processFn:[] }) builtin processors CaddyJSON(processorOptions) creates default config: CaddyJSON 1`] =
exports[`ViteCspPlugin({ processFn:[] }) builtin processors CaddyJSON(processorOptions) creates custom config: CaddyJSON 1`] = `
"{
  \\"listen\\": [
    \\":443\\"
  ],
  \\"routes\\": [
    {
      \\"match\\": [
        {
          \\"host\\": [
            \\"localhost\\"
          ]
        }
      ],
      \\"handle\\": [
        {
          \\"handler\\": \\"file_server\\",
          \\"root\\": \\"/var/www\\"
        },
        {
          \\"handler\\": \\"headers\\",
          \\"response\\": {
            \\"set\\": {
              \\"Content-Security-Policy-Report-Only\\": [
                \\"\\"
              ],
              \\"Content-Security-Policy\\": [
                \\"\\"
              ],
              \\"Report-To\\": [
                \\"\\"
              ],
              \\"Referrer-Policy\\": [
                \\"\\"
              ]
            }
          }
        }
      ]
    }
  ]
}"
`;

exports[`ViteCspPlugin({ processFn:[] }) builtin processors CaddyJSON_HeadersOnly(processorOptions) creates default config: CaddyJSON_HeadersOnly 1`] =
exports[`ViteCspPlugin({ processFn:[] }) builtin processors CaddyJSON_HeadersOnly(processorOptions) creates custom config: CaddyJSON_HeadersOnly 1`] = `
"{
  \\"handler\\": \\"headers\\",
  \\"response\\": {
    \\"set\\": {
      \\"Content-Security-Policy-Report-Only\\": [
        \\"\\"
      ],
      \\"Content-Security-Policy\\": [
        \\"\\"
      ],
      \\"Report-To\\": [
        \\"\\"
      ],
      \\"Referrer-Policy\\": [
        \\"\\"
      ]
    }
  }
}"
`;

exports[`ViteCspPlugin({ processFn:[] }) builtin processors Caddyfile(processorOptions) creates default config: Caddyfile 1`] =
exports[`ViteCspPlugin({ processFn:[] }) builtin processors Caddyfile(processorOptions) creates custom config: Caddyfile 1`] = `
"
		{$SITE_ADDRESS}

		root * dist
		header Content-Security-Policy-Report-Only \\"\\"
		header Content-Security-Policy \\"\\"
		header Report-To \\"\\"
		header Referrer-Policy \\"\\"

		file_server
		"
`;

exports[`ViteCspPlugin({ processFn:[] }) builtin processors Caddyfile_HeadersOnly(processorOptions) creates default config: Caddyfile_HeadersOnly 1`] =
exports[`ViteCspPlugin({ processFn:[] }) builtin processors Caddyfile_HeadersOnly(processorOptions) creates custom config: Caddyfile_HeadersOnly 1`] = `
"header Content-Security-Policy-Report-Only \\"\\"
header Content-Security-Policy \\"\\"
header Report-To \\"\\"
header Referrer-Policy \\"\\"
"
`;

exports[`ViteCspPlugin({ processFn:[] }) builtin processors Nginx(processorOptions) creates default config: Nginx 1`] =
exports[`ViteCspPlugin({ processFn:[] }) builtin processors Nginx(processorOptions) creates custom config: Nginx 1`] = `
"
		server {
			listen	443 80;
			index	index.html Index.html;
			add_header Content-Security-Policy-Report-Only \\"\\";
			add_header Content-Security-Policy \\"\\";
			add_header Report-To \\"\\";
			add_header Referrer-Policy \\"\\";

			location / {
				try_files	$uri /index.html $uri/ / =404;
			}
		}
		"
`;

exports[`ViteCspPlugin({ processFn:[] }) builtin processors Nginx_HeadersOnly(processorOptions) creates default config: Nginx_HeadersOnly 1`] =
exports[`ViteCspPlugin({ processFn:[] }) builtin processors Nginx_HeadersOnly(processorOptions) creates custom config: Nginx_HeadersOnly 1`] = `
"add_header Content-Security-Policy-Report-Only \\"\\";
add_header Content-Security-Policy \\"\\";
add_header Report-To \\"\\";
add_header Referrer-Policy \\"\\";
"
`;

exports[`ViteCspPlugin({ processFn:[] }) builtin processors plugin.transformIndexHtml() => buildConfigs() creates custom config for CaddyJSON: CaddyJSON 1`] =
exports[`ViteCspPlugin({ processFn:[] }) builtin processors plugin.transformIndexHtml() => buildConfigs() creates default config for CaddyJSON: CaddyJSON 1`] = `
"{
  \\"listen\\": [
    \\":443\\"
  ],
  \\"routes\\": [
    {
      \\"match\\": [
        {
          \\"host\\": [
            \\"localhost\\"
          ]
        }
      ],
      \\"handle\\": [
        {
          \\"handler\\": \\"file_server\\",
          \\"root\\": \\"/var/www\\"
        },
        {
          \\"handler\\": \\"headers\\",
          \\"response\\": {
            \\"set\\": {
              \\"Content-Security-Policy-Report-Only\\": [
                \\"\\"
              ],
              \\"Content-Security-Policy\\": [
                \\"object-src 'none'; script-src 'unsafe-inline' 'self' 'unsafe-eval'; script-src-attr; style-src 'unsafe-inline' 'self' 'unsafe-eval'; style-src-attr; base-uri 'self';\\"
              ],
              \\"Report-To\\": [
                \\"\\"
              ],
              \\"Referrer-Policy\\": [
                \\"strict-origin-when-cross-origin\\"
              ]
            }
          }
        }
      ]
    }
  ]
}"
`;

exports[`ViteCspPlugin({ processFn:[] }) builtin processors plugin.transformIndexHtml() => buildConfigs() creates custom config for CaddyJSON_HeadersOnly: CaddyJSON_HeadersOnly 1`] =
exports[`ViteCspPlugin({ processFn:[] }) builtin processors plugin.transformIndexHtml() => buildConfigs() creates default config for CaddyJSON_HeadersOnly: CaddyJSON_HeadersOnly 1`] = `
"{
  \\"handler\\": \\"headers\\",
  \\"response\\": {
    \\"set\\": {
      \\"Content-Security-Policy-Report-Only\\": [
        \\"\\"
      ],
      \\"Content-Security-Policy\\": [
        \\"object-src 'none'; script-src 'unsafe-inline' 'self' 'unsafe-eval'; script-src-attr; style-src 'unsafe-inline' 'self' 'unsafe-eval'; style-src-attr; base-uri 'self';\\"
      ],
      \\"Report-To\\": [
        \\"\\"
      ],
      \\"Referrer-Policy\\": [
        \\"strict-origin-when-cross-origin\\"
      ]
    }
  }
}"
`;

exports[`ViteCspPlugin({ processFn:[] }) builtin processors plugin.transformIndexHtml() => buildConfigs() creates custom config for Caddyfile: Caddyfile 1`] =
exports[`ViteCspPlugin({ processFn:[] }) builtin processors plugin.transformIndexHtml() => buildConfigs() creates default config for Caddyfile: Caddyfile 1`] = `
"
		{$SITE_ADDRESS}

		root * dist
		header Content-Security-Policy-Report-Only \\"\\"
		header Content-Security-Policy \\"object-src 'none'; script-src 'unsafe-inline' 'self' 'unsafe-eval'; script-src-attr; style-src 'unsafe-inline' 'self' 'unsafe-eval'; style-src-attr; base-uri 'self';\\"
		header Report-To \\"\\"
		header Referrer-Policy \\"strict-origin-when-cross-origin\\"

		file_server
		"
`;

exports[`ViteCspPlugin({ processFn:[] }) builtin processors plugin.transformIndexHtml() => buildConfigs() creates custom config for Caddyfile_HeadersOnly: Caddyfile_HeadersOnly 1`] =
exports[`ViteCspPlugin({ processFn:[] }) builtin processors plugin.transformIndexHtml() => buildConfigs() creates default config for Caddyfile_HeadersOnly: Caddyfile_HeadersOnly 1`] = `
"header Content-Security-Policy-Report-Only \\"\\"
header Content-Security-Policy \\"object-src 'none'; script-src 'unsafe-inline' 'self' 'unsafe-eval'; script-src-attr; style-src 'unsafe-inline' 'self' 'unsafe-eval'; style-src-attr; base-uri 'self';\\"
header Report-To \\"\\"
header Referrer-Policy \\"strict-origin-when-cross-origin\\"
"
`;

exports[`ViteCspPlugin({ processFn:[] }) builtin processors plugin.transformIndexHtml() => buildConfigs() creates custom config for Nginx: Nginx 1`] =
exports[`ViteCspPlugin({ processFn:[] }) builtin processors plugin.transformIndexHtml() => buildConfigs() creates default config for Nginx: Nginx 1`] = `
"
		server {
			listen	443 80;
			index	index.html Index.html;
			add_header Content-Security-Policy-Report-Only \\"\\";
			add_header Content-Security-Policy \\"object-src 'none'; script-src 'unsafe-inline' 'self' 'unsafe-eval'; script-src-attr; style-src 'unsafe-inline' 'self' 'unsafe-eval'; style-src-attr; base-uri 'self';\\";
			add_header Report-To \\"\\";
			add_header Referrer-Policy \\"strict-origin-when-cross-origin\\";

			location / {
				try_files	$uri /index.html $uri/ / =404;
			}
		}
		"
`;
exports[`ViteCspPlugin({ processFn:[] }) builtin processors plugin.transformIndexHtml() => buildConfigs() creates custom config for Nginx_HeadersOnly: Nginx_HeadersOnly 1`] =
exports[`ViteCspPlugin({ processFn:[] }) builtin processors plugin.transformIndexHtml() => buildConfigs() creates default config for Nginx_HeadersOnly: Nginx_HeadersOnly 1`] = `
"add_header Content-Security-Policy-Report-Only \\"\\";
add_header Content-Security-Policy \\"object-src 'none'; script-src 'unsafe-inline' 'self' 'unsafe-eval'; script-src-attr; style-src 'unsafe-inline' 'self' 'unsafe-eval'; style-src-attr; base-uri 'self';\\";
add_header Report-To \\"\\";
add_header Referrer-Policy \\"strict-origin-when-cross-origin\\";
"
`;
